File: server/session_handler/api_session/window_handler.py
import json
import logging
import os

import ffmpeg

from core import session_settings


class WindowHandler:
    def __get_video_byterate(self, video_path: str) -> float:
        file_size = os.path.getsize(video_path)
        probe = ffmpeg.probe(video_path)
        probe_format = probe["format"]

        if "duration" in probe_format:
            duration = float(probe_format["duration"])
        else:
            logging.error("Unable to get video duration with ffmpeg.")
            logging.error(
                "Here is the dictionary returned by ffmpeg.probe:\n%s",
                json.dumps(probe, indent=4),
            )
            raise ValueError("Unable to get video duration with ffmpeg.")
        byterate = file_size / duration
        logging.info("Byte rate calculated successfully: %f bytes per second", byterate)
        return byterate

    def __get_max_video_byterate(self, video_path: str) -> float:
        video_byterate = self.__get_video_byterate(video_path)
        return video_byterate * 1.5

    # State actions
    def __duplicate_window_size(self):
        self.__current_window_size *= 2

    def __starts_to_begging(self):
        self.__threshould = self.__current_window_size / 2
        self.__current_window_size = int(self.__video_byterate * 0.2)
        self.__current_window_size -= (
            self.__current_window_size % session_settings.cluster_size
        )

    def __duplicate_until_threshould(self):
        if self.__current_window_size * 2 > self.__threshould:
            return
        self.__current_window_size = self.__current_window_size * 2

    def __increment_slowly(self):
        new_window_size = self.__current_window_size + session_settings.cluster_size
        if new_window_size >= self.__threshould:
            return
        self.__current_window_size = new_window_size

    def __keep_current_window_size(self):
        logging.info("keeping the current window size")

    def __recover_from_loss(self):
        """
        This state try recovery from a package loss backing slowly the window size
        """
        new_window_size = self.__current_window_size - session_settings.cluster_size*2
        if new_window_size < int(self.__video_byterate * 0.2):
            logging.info("window already reached to inferior threshould, nothing to do")
            return
        self.__threshould = new_window_size + session_settings.cluster_size
        self.__current_window_size = new_window_size

    # conditional transitions
    def __check_loss_percentage(self, loss_percentage: float):
        return loss_percentage-session_settings.at_most_loss_percentage >= 1e-5

    def __state_zero_transitions(self, loss_percentage: float):

        if self.__check_loss_percentage(loss_percentage):
            self.__current_state = 1
            return
        if self.__current_window_size * 2 >= self.__threshould:
            self.__current_state = 3
            return
        self.__current_state = 0

    def __state_one_transitions(self, loss_percentage: float):
        self.__current_state = 1 if self.__check_loss_percentage(loss_percentage) else 2

    def __state_two_transitions(self, loss_percentage: float):
        if self.__check_loss_percentage(loss_percentage):
            self.__current_state = 1
            return
        if self.__current_window_size * 2 >= self.__threshould:
            self.__current_state = 3
            return
        self.__current_state = 2

    def __state_three_transitions(self, loss_percentage: float):
        if self.__check_loss_percentage(loss_percentage):
            self.__current_state = 1
            return
        if (
            self.__current_window_size + session_settings.cluster_size
            >= self.__threshould
        ):
            self.__current_state = 4
            return
        self.__current_state = 3

    def __state_four_transitions(self, loss_percentage: float):
        if not self.__check_loss_percentage(loss_percentage):
            self.__current_state = 4
            return
        self.__current_state = 5

    def __state_five_transitions(self, loss_percentage:float):
        if self.__check_loss_percentage(loss_percentage):
            self.__current_state = 5
            return
        self.__current_state = 3

    def __init__(self, video_path: str) -> None:
        # Ensure the window size is a multiple of the OS cluster size
        # to avoid reading unnecessary disk blocks and maximize data usage efficiency.
        video_byterate = self.__get_max_video_byterate(video_path)
        self.__video_byterate = video_byterate
        self.__current_window_size = int(video_byterate * 0.2)
        self.__current_window_size -= (
            self.__current_window_size % session_settings.cluster_size
        )
        self.__threshould = video_byterate
        self.__current_state = 0
        self.__states_actions_map = [
            self.__duplicate_window_size,
            self.__starts_to_begging,
            self.__duplicate_until_threshould,
            self.__increment_slowly,
            self.__keep_current_window_size,
            self.__recover_from_loss
        ]
        self.__state_transitions_map = [
            self.__state_zero_transitions,
            self.__state_one_transitions,
            self.__state_two_transitions,
            self.__state_three_transitions,
            self.__state_four_transitions,
            self.__state_five_transitions
        ]

    def update_window_size(self, byte_count: int):
        logging.info("throughput: %d, bytes losed: %d", byte_count, self.__current_window_size-byte_count)
        loss_percentage = 1.0 - byte_count / self.__current_window_size
        self.__state_transitions_map[self.__current_state](loss_percentage)
        self.__states_actions_map[self.__current_state]()
        logging.info("update window size to %d", self.__current_window_size)
        logging.info("current state: %d", self.__current_state)

    def get_window_size(self) -> int:
        assert self.__current_window_size % session_settings.cluster_size == 0
        return self.__current_window_size

File: serverjava/src/session_handler/apiSession/WindowHandler.java
package session_handler.apiSession;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.logging.Logger;
import session_handler.core.Settings;

public class WindowHandler {

    private static final Logger logger = Logger.getLogger(WindowHandler.class.getName());
    private double videoByterate;
    private int currentWindowSize;
    private double threshold;
    private int currentState;
    private final int clusterSize;

    private enum State {
        STATE_ZERO,
        STATE_ONE,
        STATE_TWO,
        STATE_THREE,
        STATE_FOUR
    }

    private final StateAction[] stateActions;
    private final StateTransition[] stateTransitions;

    public WindowHandler(String videoPath) throws IOException {
        // Initialize clusterSize
        this.clusterSize = Settings.SessionSettings.getClusterSize();

        // Initialize other parameters
        this.videoByterate = getVideoByterate(videoPath);
        this.currentWindowSize = (int) (videoByterate * 0.05);
        this.currentWindowSize -= (this.currentWindowSize % clusterSize);
        this.threshold = videoByterate;
        this.currentState = State.STATE_ZERO.ordinal();

        // Initialize state actions and transitions
        stateActions = new StateAction[]{
            this::duplicateWindowSize,
            this::startsToBegging,
            this::duplicateUntilThreshold,
            this::incrementSlowly,
            this::keepCurrentWindowSize
        };

        stateTransitions = new StateTransition[]{
            this::stateZeroTransitions,
            this::stateOneTransitions,
            this::stateTwoTransitions,
            this::stateThreeTransitions,
            this::stateFourTransitions
        };
    }

    private double getVideoByterate(String videoPath) throws IOException {
        File file = new File(videoPath);
        double duration = getVideoDurationWithFFmpeg(file);
        if (duration > 0) {
            double byterate = (double) file.length() / duration;
            logger.info(String.format("Byte rate calculated successfully: %f bytes per second", byterate));
            return byterate;
        } else {
            logger.severe("Unable to calculate video byte rate with FFmpeg.");
            throw new IOException("Unable to calculate video byte rate with FFmpeg.");
        }
    }

    private double getVideoDurationWithFFmpeg(File file) throws IOException {
        ProcessBuilder processBuilder = new ProcessBuilder("ffmpeg", "-i", file.getAbsolutePath());
        processBuilder.redirectErrorStream(true);

        Process process = processBuilder.start();
        InputStream inputStream = process.getInputStream();

        try (java.util.Scanner scanner = new java.util.Scanner(inputStream).useDelimiter("\\A")) {
            String output = scanner.hasNext() ? scanner.next() : "";

            String durationLine = output.lines()
                .filter(line -> line.contains("Duration:"))
                .findFirst()
                .orElse("");

            String[] parts = durationLine.split("Duration:")[1].trim().split(",")[0].trim().split(":");
            if (parts.length == 3) {
                double hours = Double.parseDouble(parts[0]);
                double minutes = Double.parseDouble(parts[1]);
                double seconds = Double.parseDouble(parts[2].split("\\.")[0]); // Remove milliseconds
                double duration = hours * 3600 + minutes * 60 + seconds;
                return duration;
            } else {
                logger.severe("Unable to parse video duration from FFmpeg output.");
                throw new IOException("Unable to parse video duration from FFmpeg output.");
            }
        } catch (Exception e) {
            logger.severe("Error reading FFmpeg output.");
            throw new IOException("Error reading FFmpeg output.", e);
        }
    }

    private void duplicateWindowSize() {
        this.currentWindowSize *= 2;
    }

    private void startsToBegging() {
        this.threshold = this.currentWindowSize / 2.0;
        this.currentWindowSize = (int) (this.videoByterate * 0.05);
        this.currentWindowSize -= (this.currentWindowSize % clusterSize);
    }

    private void duplicateUntilThreshold() {
        if (this.currentWindowSize * 2 > this.threshold) {
            return;
        }
        this.currentWindowSize *= 2;
    }

    private void incrementSlowly() {
        int newWindowSize = this.currentWindowSize + clusterSize;
        if (newWindowSize >= this.threshold) {
            return;
        }
        this.currentWindowSize = newWindowSize;
    }

    private void keepCurrentWindowSize() {
        logger.info("Keeping the current window size");
    }

    private void stateZeroTransitions(boolean lossesOccurred) {
        if (this.currentWindowSize * 2 >= this.threshold) {
            this.currentState = State.STATE_THREE.ordinal();
        } else {
            this.currentState = lossesOccurred ? State.STATE_ONE.ordinal() : State.STATE_ZERO.ordinal();
        }
    }

    private void stateOneTransitions(boolean lossesOccurred) {
        this.currentState = lossesOccurred ? State.STATE_ONE.ordinal() : State.STATE_TWO.ordinal();
    }

    private void stateTwoTransitions(boolean lossesOccurred) {
        if (lossesOccurred) {
            this.currentState = State.STATE_ONE.ordinal();
        } else if (this.currentWindowSize * 2 >= this.threshold) {
            this.currentState = State.STATE_THREE.ordinal();
        } else {
            this.currentState = State.STATE_TWO.ordinal();
        }
    }

    private void stateThreeTransitions(boolean lossesOccurred) {
        if (lossesOccurred) {
            this.currentState = State.STATE_ONE.ordinal();
        } else if (this.currentWindowSize + clusterSize >= this.threshold) {
            this.currentState = State.STATE_FOUR.ordinal();
        } else {
            this.currentState = State.STATE_THREE.ordinal();
        }
    }

    private void stateFourTransitions(boolean lossesOccurred) {
        if (!lossesOccurred) {
            this.currentState = State.STATE_FOUR.ordinal();
        } else {
            this.currentState = State.STATE_ZERO.ordinal();
            this.threshold = this.videoByterate;
            this.currentWindowSize = (int) (this.videoByterate * 0.05);
            this.currentWindowSize -= (this.currentWindowSize % clusterSize);
        }
    }

    public void updateWindowSize(int byteCount) {
        boolean lossesOccurred = (byteCount / this.currentWindowSize) != 1;
        stateTransitions[this.currentState].transition(lossesOccurred);
        stateActions[this.currentState].action();
        logger.info(String.format("Update window size to %d", this.currentWindowSize));
        logger.info(String.format("Current state: %d", this.currentState));
    }

    public int getWindowSize() {
        assert this.currentWindowSize % clusterSize == 0;
        return this.currentWindowSize;
    }

    @FunctionalInterface
    private interface StateAction {
        void action();
    }

    @FunctionalInterface
    private interface StateTransition {
        void transition(boolean lossesOccurred);
    }
}